# Taking Input in Python: input(), Data Conversion & Internal Execution

1. `Read User Input` - `input()` waits for the user to type and press enter.

2. `Standard Input(stdin)` : The input is taken as `string` from the system's `stdin buffer`.

3. `Memory Allocation` : The entered strings is stored in memory until assigned to a variable.

4. `Blocking Operations` : Exectuion halts until input is provided.

5. `Optimization Tip` - Converts input to the require type (`int(input())` ti avoid type issues.

# Syntax & Structure
### Why is Understanding `input()` important?

- Helps build `interactive programs` (CLI applications, chatbots,user-driven scripts)

- Avoid `data type mismatches` by properly handling conversions.

- `Optimizes `user input handling` for better performance and security.


###

In [2]:
# Basic Syntax of input()

variable=input("promt message")

## Explanation

- `variable` : Stores the user's input.
- `input("Prmot message")`: Displays a message and `wait for the user` to enter a value.
- `The `input always stored as a string` even if the users enters a numbers.

In [6]:
# Example of caputring user input
name=input("enter your name:")
age=input("enter your age:")
print("Hello,",name,"Your are",age,"years old")

Hello, raghav Your are 21 years old


## How `input()` works internally
### Breaking Down `input()` Execution

1. `Pauses program execution` and waits for the user to type input.

2. `Read input as a string`(even if the users enters number)

3. `Stores the input in memory` as a dynamically allocated object.

4. `Returns the input as a strings`, which can be further processed.

In [10]:
user_input=input("enter something")
print("You entered:",user_input)
print(type(user_input))

You entered: 123
<class 'str'>


- Even if the user enters `123`,it will be stored as `"123"`(string).

- To use it as an `integer`,explicit conversion required.(`int(user_input`))

In [ ]:
## Handling Different Input Data Types
### Converting Input to Other Data Types

num = int(input("Enter a number: "))  # Converts input string to an integer
pi = float(input("Enter a decimal number: "))  # Converts input string to float
is_active = bool(int(input("Enter 1 for True, 0 for False: ")))  # Converts input to boolean

print(num, pi, is_active)
print(num)
print(pi)
print(is_active)


1 2.0 True
1
2.0
True


## Key Takaways

- `int()` : Converts input to an `integer`.
- `float()` : Converts input a `floating point number`.
- `bool(int())` : Converts `1 to True and 0 to False`.

In [ ]:
# Handling Multiple Inputs at Once
x,y=map(int,input("enter two number seperated by 't':").split('t'))
print(x,y)
print("sum", x+y)




10 20
sum 30


# How it Work?

- `split()` breakes the input string into a list.
- `map(int,...)` converts each value to an integer.


## Handline Edge Cases in `input()`

### Why Handling Input Errors is Important?

- Users inputs are `unpredictable` to `prevent creashes`, error handling is essential.

In [34]:
# Handling Invalid Data(try-except)
try:
    age=int(input("enter your age:"))
    print("your age is",age)
except ValueError:
    print("Invalid input!please enter a number")

Invalid input!please enter a number


# Example Output
```
Enter your age: twenty
Invalid input! Please enter a number.
```


- Prevent crashes when a user enters `non-numeric input`.

In [38]:
# Providing Default Values
name=input("enter your name") or "people"
print("Welcome,",name)

Welcome, people


- `Example Input` : (User presses enter without typing anything)

```
press(enter)
```

- Output
```
Welcome,people
```

- How it Works?

    - `or people` ensures a default value if the user enter nothing.

## Advanced Techniques For Taking Input

- Reading a input without a Newline (`sys.stdin.read())` useful when handling `large input at once`(like reading from files).

- Unlike `input()`,this read all available `input until EOF(End of File).`

# Using `getpass()` for Secure input(Passowrds)

In [3]:
from getpass import getpass
password=getpass("Enter your password")
print(password)
print("Password entered(hidden for security)")

23456
Password entered(hidden for security)


# Why Use `getpass()` Instead of `input()`?

- `Prevent passwords from being displayed on the screen`.
- `safer than using input() for sensitive data`.


## Optimize User Input Handling for Performance
### Why Optimize Input Handling?

- Handling large amounts of input efficiently imporves performance.

- Essential for `competitive programming,data processing,and real time application`.

In [ ]:
import sys
num = int(sys.stdin.readline().strip())  # Faster than input() for large input
print("You entered:", num)

# Why `sys.stdin.readline()`?

- `~10x faster` than `input()` for reading large data points.
- Remove trailing newline with `.strip()`.

# Key Takeways

- `input()` always returns a string requiring `type conversion` when necessary.
- Use `try-except` to prevent crashes form invalid user input.
- `Multiple values` can be taken using `split` and `map()`
- Use `getpass()` for secure password entry.
- For `large input`,use `sys.stdin.readline()` for better performance.

In [ ]:
import sys

lines = []
for line in sys.stdin:
    line = line.strip()
    if line:
        lines.append(line)
        print("You entered:", line)

# Optional: Print all inputs at the end
print("\nAll Inputs:")
for line in lines:
    print(line)